src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java

```diff
@@ -17,156 +17,830 @@
+            }
+        }
+
+        TableSnapshot snapshot = new TableSnapshot(cfs.metadata.keyspace,
+                                                   cfs.metadata.name,
+                                                   cfs.metadata.id.asUUID(),
+                                                   tag,
+                                                   creationTime,
+                                                   SnapshotManifest.computeExpiration(ttl, creationTime),
+                                                   snapshotDirs,
+                                                   ephemeral);
+
+        addSnapshot(snapshot);
+        return snapshot;
     }
 
-    @VisibleForTesting
-    public static void shutdownAndWait(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException
+    // MBean methods
+
+    @Override
+    public void takeSnapshot(String tag, Map<String, String> options, String... entities) throws IOException
     {
-        ExecutorUtils.shutdownNowAndWait(timeout, unit, executor);
+        if (StorageService.instance.operationMode() == StorageService.Mode.JOINING)
+            throw new IOException("Cannot snapshot until bootstrap completes");
+
+        if (tag == null || tag.isEmpty())
+            throw new IOException("You must supply a snapshot name.");
+
+        DurationSpec.IntSecondsBound ttl = options.containsKey("ttl") ? new DurationSpec.IntSecondsBound(options.get("ttl")) : null;
+        if (ttl != null)
+        {
+            int minAllowedTtlSecs = CassandraRelevantProperties.SNAPSHOT_MIN_ALLOWED_TTL_SECONDS.getInt();
+            if (ttl.toSeconds() < minAllowedTtlSecs)
+                throw new IllegalArgumentException(format("ttl for snapshot must be at least %d seconds", minAllowedTtlSecs));
+        }
+
+        boolean skipFlush = Boolean.parseBoolean(options.getOrDefault("skipFlush", "false"));
+
+        Map<Keyspace, Set<ColumnFamilyStore>> entitiesForSnapshot = parseEntitiesForSnapshot(entities);
+
+        for (Map.Entry<Keyspace, Set<ColumnFamilyStore>> entry : entitiesForSnapshot.entrySet())
+        {
+            for (ColumnFamilyStore table : entry.getValue())
+            {
+                String keyspaceName = table.getKeyspaceName();
+                String tableName = table.getTableName();
+                if (getSnapshot(table.getKeyspaceName(), table.getTableName(), tag).isPresent())
+                    throw new IOException(format("Snapshot %s for %s.%s already exists.", tag, keyspaceName, tableName));
+            }
+        }
+
+        Instant creationTime = now();
+
+        for (Map.Entry<Keyspace, Set<ColumnFamilyStore>> entry : entitiesForSnapshot.entrySet())
+        {
+            Keyspace keyspace = entry.getKey();
+            for (ColumnFamilyStore table : entry.getValue())
+                keyspace.snapshot(tag, table.getTableName(), skipFlush, ttl, snapshotRateLimiter, creationTime);
```

Comment:
We already have all the `ColumnFamilyStore` objects we want to snapshot here. Why can't we bypass `Keyspace` altogether and just call `ColumnFamilyStore.snapshot` for each snapshot in the `entitiesForSnapshot` map ?

I think this would allow removing or simplifying most snapshot logic from [Keyspace](https://github.com/apache/cassandra/blob/fe025c7f79e76d99e0db347518a7872fd4a114bc/src/java/org/apache/cassandra/db/Keyspace.java#L239) and  ColumnFamilyStore which is ultimately calling [SnapshotManager.instance.createSnapshot](https://github.com/instaclustr/cassandra/blob/CASSANDRA-18111/src/java/org/apache/cassandra/db/ColumnFamilyStore.java#L2138).

Lmk if this makes sense.